### ISSUE TITLE
SSE-C key does not support base64, only ASCII printable characters

### ISSUE THREAD

**Issue by nreynis** [minio/minio#7783]
It seems that SSE-C keys do not support base64 encoded keys, despite the [documentation](https://docs.min.io/docs/how-to-use-minio-s-server-side-encryption-with-aws-cli.html) stating the contrary.

I'm using the official minio docker container.

When using the javascript aws-sdk I got 403 errors. If I use a plaintext key instead of a base64 encoded one, everythings works fine.

I also reproduce the problem with the official mc client:
```sh
MC_ENCRYPT_KEY=ytest/test-bucket=$(< .secrets/minio_crypt.key)
mc cp --insecure --encrypt-key "$MC_ENCRYPT_KEY" test1.txt ytest/test-bucket
mc: <ERROR> Unable to parse encryption keys. SSE-C prefix=secret should be delimited by , and secret should be 32 bytes long.
```
Everything works fine if my key is exactly 32 bits long and use only ASCII printable characters.

This is obviously a problem because it reduce the effective entropy of the key to 12bits `Math.ceil((126-31)/255*32)` instead of  32, which is really weak.

Am I doing something wrong? Can I achieve real 32 bits entropy ?

**Comment by harshavardhana** [comment]
Closing as duplicate of https://github.com/minio/mc/issues/2678

**Comment by nreynis** [comment]
@harshavardhana While the issue seems to overlap it's not limited to the minio/mc client. The server component is also concerned.

**Comment by harshavardhana** [comment]
Server implementation is correct in accordance with AWS S3 spec.

**Comment by harshavardhana** [comment]
In your case the issue is because mc internally converts the key into base64, it doesn't expect you to do base64 prior.

**Comment by nreynis** [comment]
I can reproduce the issue by using plain old javascript with `aws-sdk`

**Comment by kannappanr** [comment]
@nreynis I am not sure how `minio` has an issue here. Can you give me a sample code and why you think this is server side bug

**Comment by harshavardhana** [comment]
> I can reproduce the issue by using plain old javascript with `aws-sdk`

As shown here https://github.com/minio/mc/issues/2678#issuecomment-501878723 the server component is behaving correctly. Just use https://play.minio.io:9000 @nreynis would be easier to reproduce.

**Comment by nreynis** [comment]
Thank you for your help, I don't have the code right here to compare with your example. I'll look at this tomorrow and keep you updated.

**Comment by kannappanr** [comment]
Here is an example that I have used 

```
$ cat /tmp/key
01234567890123456789012345678901
```
```
$ base64 /tmp/key
MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDE=
```

```
package main

import (
	"fmt"
	"io/ioutil"
	"os"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
)

func setupClient() *s3.S3 {

	accessKey := "Q3AM3UQ867SPQQA43P2F"
	secretKey := "zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG"
	sdkEndpoint := "https://play.minio.io:9000"

	creds := credentials.NewStaticCredentials(accessKey, secretKey, "")
	newSession := session.New()
	s3Config := &aws.Config{
		Credentials:      creds,
		Endpoint:         aws.String(sdkEndpoint),
		Region:           aws.String("us-east-1"),
		S3ForcePathStyle: aws.Bool(true),
		DisableSSL:       aws.Bool(false),
	}

	return s3.New(newSession, s3Config)
}
func main() {

	s3Client := setupClient()

	if s3Client == nil {
		fmt.Println("s3Client is nil, connection failed")
		return
	}
	fmt.Println("1")
	bucket := "test"
	object := "ssecsingle"
	f, err := os.Open("/etc/issue")
	if err != nil {
		fmt.Println(err)
	}

	op, err := s3Client.PutObject(&s3.PutObjectInput{
		Bucket:               aws.String(bucket),
		Key:                  aws.String(object),
		Body:                 f,
		SSECustomerAlgorithm: aws.String("AES256"),
		SSECustomerKey:       aws.String("01234567890123456789012345678901"),
	})
	fmt.Println(op)
	objectoutput, err := s3Client.GetObject(&s3.GetObjectInput{
		Bucket:               aws.String(bucket),
		Key:                  aws.String(object),
		SSECustomerAlgorithm: aws.String("AES256"),
		SSECustomerKey:       aws.String("01234567890123456789012345678901"),
	})

	if err != nil {
		fmt.Println("Error1", err)
		return
	}

	b, err := ioutil.ReadAll(objectoutput.Body)
	fmt.Println("body", string(b), err)
}
```
Here is the trace on the server side

```
127.0.0.1 [REQUEST objectAPIHandlers.PutObject] [21:29:03.601806656]
127.0.0.1 PUT /test/ssecsingle
127.0.0.1 Host: play.minio.io:9000
127.0.0.1 X-Amz-Date: 20190613T212903Z
127.0.0.1 X-Amz-Server-Side-Encryption-Customer-Key: MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDE=
127.0.0.1 X-Amz-Server-Side-Encryption-Customer-Key-Md5: KYvwGXoFFJ42a2u2GDWhwQ==
127.0.0.1 User-Agent: aws-sdk-go/1.20.0 (go1.12.1; linux; amd64)
127.0.0.1 X-Amz-Content-Sha256: 1094a1ff5d6d0d9137ad321396fe5c2a4653258b102f9ac04ec7acae00970598
127.0.0.1 Content-Length: 26
127.0.0.1 Content-Md5: uVRBjmpQ1NTLjwJ3bYZ1UA==
127.0.0.1 X-Amz-Server-Side-Encryption-Customer-Algorithm: AES256
127.0.0.1 Accept-Encoding: gzip
127.0.0.1 Authorization: AWS4-HMAC-SHA256 Credential=Q3AM3UQ867SPQQA43P2F/20190613/us-east-1/s3/aws4_request, SignedHeaders=content-length;content-md5;host;x-amz-content-sha256;x-amz-date;x-amz-server-side-encryption-customer-algorithm;x-amz-server-side-encryption-customer-key;x-amz-server-side-encryption-customer-key-md5, Signature=f8fefc97698f3fe040f212bf9a7caebdf37d60ba1b072403d2d9724374dfbdae
127.0.0.1 <BODY>
127.0.0.1 [RESPONSE] [21:29:03.611452560]
127.0.0.1 200 OK
127.0.0.1 X-Amz-Server-Side-Encryption-Customer-Algorithm: AES256
127.0.0.1 X-Amz-Server-Side-Encryption-Customer-Key-Md5: KYvwGXoFFJ42a2u2GDWhwQ==
127.0.0.1 X-Minio-Deployment-Id: 41e39f4a-3b66-415b-9ddf-025d76a58668
127.0.0.1 X-Xss-Protection: 1; mode=block
127.0.0.1 Content-Security-Policy: block-all-mixed-content
127.0.0.1 ETag: "89eaca4ccea0f4554c51c3eb239ff811"
127.0.0.1 Server: MinIO/DEVELOPMENT.2019-06-13T20-58-18Z
127.0.0.1 X-Amz-Bucket-Region: us-east-1
127.0.0.1 Accept-Ranges: bytes
127.0.0.1 Content-Length: 0
127.0.0.1 Vary: Origin
127.0.0.1 X-Amz-Request-Id: 15A7DF5307114E41
127.0.0.1 <BODY>127.0.0.1 
127.0.0.1 [REQUEST objectAPIHandlers.GetObject] [21:29:03.620631173]
127.0.0.1 GET /test/ssecsingle
127.0.0.1 Host: play.minio.io:9000
127.0.0.1 Accept-Encoding: gzip
127.0.0.1 User-Agent: aws-sdk-go/1.20.0 (go1.12.1; linux; amd64)
127.0.0.1 X-Amz-Server-Side-Encryption-Customer-Key: MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDE=
127.0.0.1 X-Amz-Server-Side-Encryption-Customer-Key-Md5: KYvwGXoFFJ42a2u2GDWhwQ==
127.0.0.1 X-Amz-Server-Side-Encryption-Customer-Algorithm: AES256
127.0.0.1 Authorization: AWS4-HMAC-SHA256 Credential=Q3AM3UQ867SPQQA43P2F/20190613/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date;x-amz-server-side-encryption-customer-algorithm;x-amz-server-side-encryption-customer-key;x-amz-server-side-encryption-customer-key-md5, Signature=0abaa9abd6f9eb63cd78d793244d05b00c27a36b1998cec551e159cf721863d7
127.0.0.1 Content-Length: 0
127.0.0.1 X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
127.0.0.1 X-Amz-Date: 20190613T212903Z
127.0.0.1 <BODY>
127.0.0.1 [RESPONSE] [21:29:03.626641173]
127.0.0.1 200 OK
127.0.0.1 Content-Length: 26
127.0.0.1 ETag: "89eaca4ccea0f4554c51c3eb239ff811"
127.0.0.1 Last-Modified: Thu, 13 Jun 2019 21:29:03 GMT
127.0.0.1 X-Amz-Bucket-Region: us-east-1
127.0.0.1 X-Amz-Server-Side-Encryption-Customer-Key-Md5: KYvwGXoFFJ42a2u2GDWhwQ==
127.0.0.1 X-Xss-Protection: 1; mode=block
127.0.0.1 Content-Type: application/octet-stream
127.0.0.1 Server: MinIO/DEVELOPMENT.2019-06-13T20-58-18Z
127.0.0.1 Vary: Origin
127.0.0.1 X-Amz-Request-Id: 15A7DF530838F3D3
127.0.0.1 X-Amz-Server-Side-Encryption-Customer-Algorithm: AES256
127.0.0.1 Accept-Ranges: bytes
127.0.0.1 Content-Security-Policy: block-all-mixed-content
127.0.0.1 X-Minio-Deployment-Id: 41e39f4a-3b66-415b-9ddf-025d76a58668
127.0.0.1 <BODY>127.0.0.1 
```
See that the header `X-Amz-Server-Side-Encryption-Customer-Key` matches the base64 encoded value.

Now, if I use the same key in `mc`, you can see that this works. 

```
$ mc cat --debug --encrypt-key "play/test=01234567890123456789012345678901" play/test/ssecsingle 
mc: <DEBUG> GET /test/?location= HTTP/1.1
Host: play.minio.io:9000
User-Agent: MinIO (linux; amd64) minio-go/v6.0.28 mc/2019-06-13T21:26:36Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20190613/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: UNSIGNED-PAYLOAD
X-Amz-Date: 20190613T214104Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Content-Length: 137
Accept-Ranges: bytes
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Date: Thu, 13 Jun 2019 21:41:04 GMT
Server: MinIO/DEVELOPMENT.2019-06-13T20-58-18Z
Vary: Origin
X-Amz-Bucket-Region: us-east-1
X-Amz-Request-Id: 15A7DFFADC5EFA9A
X-Minio-Deployment-Id: 41e39f4a-3b66-415b-9ddf-025d76a58668
X-Xss-Protection: 1; mode=block

mc: <DEBUG> Response Time:  56.537964ms

mc: <DEBUG> GET /test/?delimiter=%2F&max-keys=1000&prefix=ssecsingle HTTP/1.1
Host: play.minio.io:9000
User-Agent: MinIO (linux; amd64) minio-go/v6.0.28 mc/2019-06-13T21:26:36Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20190613/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20190613T214104Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Content-Length: 581
Accept-Ranges: bytes
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Date: Thu, 13 Jun 2019 21:41:04 GMT
Server: MinIO/DEVELOPMENT.2019-06-13T20-58-18Z
Vary: Origin
X-Amz-Bucket-Region: us-east-1
X-Amz-Request-Id: 15A7DFFADD33AD52
X-Minio-Deployment-Id: 41e39f4a-3b66-415b-9ddf-025d76a58668
X-Xss-Protection: 1; mode=block

mc: <DEBUG> Response Time:  12.816401ms

mc: <DEBUG> GET /test/ssecsingle HTTP/1.1
Host: play.minio.io:9000
User-Agent: MinIO (linux; amd64) minio-go/v6.0.28 mc/2019-06-13T21:26:36Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20190613/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date;x-amz-server-side-encryption-customer-algorithm;x-amz-server-side-encryption-customer-key;x-amz-server-side-encryption-customer-key-md5, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20190613T214104Z
X-Amz-Server-Side-Encryption-Customer-Algorithm: AES256
X-Amz-Server-Side-Encryption-Customer-Key: MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDE=
X-Amz-Server-Side-Encryption-Customer-Key-Md5: KYvwGXoFFJ42a2u2GDWhwQ==
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Content-Length: 26
Accept-Ranges: bytes
Content-Security-Policy: block-all-mixed-content
Content-Type: application/octet-stream
Date: Thu, 13 Jun 2019 21:41:04 GMT
Etag: "89eaca4ccea0f4554c51c3eb239ff811"
Last-Modified: Thu, 13 Jun 2019 21:29:03 GMT
Server: MinIO/DEVELOPMENT.2019-06-13T20-58-18Z
Vary: Origin
X-Amz-Bucket-Region: us-east-1
X-Amz-Request-Id: 15A7DFFADE11FF20
X-Amz-Server-Side-Encryption-Customer-Algorithm: AES256
X-Amz-Server-Side-Encryption-Customer-Key-Md5: KYvwGXoFFJ42a2u2GDWhwQ==
X-Minio-Deployment-Id: 41e39f4a-3b66-415b-9ddf-025d76a58668
X-Xss-Protection: 1; mode=block

mc: <DEBUG> Response Time:  32.656707ms

Ubuntu 18.04.2 LTS \n \l
```

**Comment by nreynis** [comment]
@harshavardhana @kannappanr 
I've jumped back to my code and found out what was wrong, I've got a double base64 encoding going on.

The aws-sdk was reencoding all strings to base64. I initially couldn't find the documentation for `putObject` params so I relied only on the typescript definitions. After some search I finally found out where the params where described:

https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#putObject-property

What really confused me was that the `X-Amz-Server-Side-Encryption-Customer-Key-Md5` wasn't behaving the same way. By digging out in the aws-sdk source code I found out that it totally ignored the value we gave him and recomputed the md5 from the original key, which is not what the documentation says.

So this was a conjonction of improper assumption on my part and wrong documentation on Amazon part: everything is fine on minio server side.

Thank you for your help.
